File: bridge/intent_router.py
"""
bridge/intent_router.py — 轻量意图识别模块

Three-layer intent classification for chat completions:
  Layer 1 (this module): keyword/regex rules (millisecond latency)
  Layer 2: (reserved for ML classifier)
  Layer 3: implicit — let the LLM handle naturally, don't infer from response

Thread-safe, stateless, no external dependencies.
"""

import re
from typing import Final


class IntentRouter:
    """Stateless intent classifier using keyword and regex matching.

    Thread-safe — no mutable instance state. All data is module-level.
    """

    # Task-action keywords — if ANY of these appear, it's a task
    _TASK_KEYWORDS: Final[list[str]] = [
        "帮我", "创建", "做一下", "跑", "查一下",
        "计算", "分析", "汇总", "生成", "写一份",
        "安排", "建个", "帮我去",
    ]

    # Single action characters specifically called out as task triggers
    _TASK_CHARS: Final[set[str]] = {"做", "跑"}

    # Chat/greeting / casual topic keywords
    _CHAT_KEYWORDS: Final[list[str]] = [
        "你好", "早安", "晚安", "哈哈", "吃",
        "睡", "笑", "开心", "好吗", "怎么样",
        "你觉得", "想", "喜欢",
    ]

    # Question indicators (Chinese + general interrogatives)
    _QUESTION_RE: Final[re.Pattern] = re.compile(
        r"[？?吗嘛吧呢]|怎么|什么|哪|谁|啥|何|"
        r"是否|有没有|能不能|会不会|要不要"
    )

    @classmethod
    def classify(cls, text: str) -> str:
        """Classify user input as ``"task"``, ``"chat"``, or ``"ambiguous"``.

        Args:
            text: Raw user input string (Chinese or mixed language).

        Returns:
            One of ``"task"``, ``"chat"``, ``"ambiguous"``.
        """
        if not text or not text.strip():
            return "ambiguous"

        stripped = text.strip()

        # ---- Rule 1: Task keywords take priority ----
        for kw in cls._TASK_KEYWORDS:
            if kw in stripped:
                return "task"
        for ch in cls._TASK_CHARS:
            if ch in stripped:
                return "task"

        # ---- Rule 2: Chat/greeting keywords ----
        for kw in cls._CHAT_KEYWORDS:
            if kw in stripped:
                return "chat"

        # ---- Rule 3: Question without action words → chat (LLM decides) ----
        if cls._QUESTION_RE.search(stripped):
            return "chat"

        # ---- Rule 4: No match → ambiguous (Layer 3) ----
        return "ambiguous"

File: bridge-rs/src/intent_router.rs
//! Intent Router — keyword/regex classifier
//! 0-risk port: pure logic, no IO. Returns "task" / "chat" / "ambiguous".

use once_cell::sync::Lazy;
use regex::Regex;

const TASK_KEYWORDS: &[&str] = &[
    "帮我", "创建", "做一下", "跑", "查一下",
    "计算", "分析", "汇总", "生成", "写一份",
    "安排", "建个", "帮我去",
];
const TASK_CHARS: &[char] = &['做', '跑'];
const CHAT_KEYWORDS: &[&str] = &[
    "你好", "早安", "晚安", "哈哈", "吃",
    "睡", "笑", "开心", "好吗", "怎么样",
    "你觉得", "想", "喜欢",
];

static QUESTION_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"[？?吗嘛吧呢]|怎么|什么|哪|谁|啥|何|是否|有没有|能不能|会不会|要不要").unwrap()
});

pub fn classify(text: &str) -> &'static str {
    if text.trim().is_empty() {
        return "ambiguous";
    }
    let stripped = text.trim();

    // Rule 1: task keywords take priority
    for kw in TASK_KEYWORDS {
        if stripped.contains(kw) {
            return "task";
        }
    }
    for ch in TASK_CHARS {
        if stripped.contains(*ch) {
            return "task";
        }
    }

    // Rule 2: chat/greeting keywords
    for kw in CHAT_KEYWORDS {
        if stripped.contains(kw) {
            return "chat";
        }
    }

    // Rule 3: question without action words → chat
    if QUESTION_RE.is_match(stripped) {
        return "chat";
    }

    // Rule 4: ambiguous (LLM decides)
    "ambiguous"
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_task() {
        assert_eq!(classify("帮我创建一个文件"), "task");
        assert_eq!(classify("跑一下测试"), "task");
    }

    #[test]
    fn test_chat() {
        assert_eq!(classify("你好"), "chat");
        assert_eq!(classify("今天怎么样？"), "chat");
    }

    #[test]
    fn test_ambiguous() {
        assert_eq!(classify(""), "ambiguous");
        assert_eq!(classify("伊卡洛斯"), "ambiguous");
    }
}
